# RNA-seq in iPSC-derived NPCs and neurons

Here we compare the RNA-seq from our iPSC-derived neurons with external brain datasets.

```{r Setup, message=FALSE, warning=FALSE, echo=FALSE}
library(dplyr)
library(readr)
library(magrittr)
library(ggplot2)
library(pheatmap)
library(pcaMethods)
library(DESeq2)

root = "/Users/jeremys/work/opentargets"
outputPath = "/Users/jeremys/work/opentargets/ipsneurons/GRCh38/RNA/analysis"
#setwd("/Users/jeremys/work/opentargets/ipsneurons/src/rna/")
gtex.rpkmfile = file.path("/Users/jeremys/work/sensoryneurons/snqtl/results/MDS", "rpkms.gtex.ips.drg.sn.rds")
gtex.metafile = file.path("/Users/jeremys/work/sensoryneurons/snqtl/results/MDS", "rpkms.sample_meta.txt")
counts.fname = file.path(outputPath, "ipsneurons.counts.txt.gz")
txRefPath = file.path(root, "reference/GRCh38/Homo_sapiens.GRCh38.89.transcript_data.rds")

saveFiles = T

source(file.path(root, "src/R/clustering.R"))
```

```{r LoadCounts, warning=FALSE, message=FALSE, echo=FALSE}
readtsv = function(fname, ...) { readr::read_tsv(fname, ...) %>% as.data.frame() }

sample.meta = readtsv(file.path(root, "ipsneurons/GRCh38/RNA/meta/ipsneurons.rna.metadata.txt"))
sample.meta$sampleID = tolower(sample.meta$sampleID)
rownames(sample.meta) = sample.meta$sampleID
sample.meta$day = as.factor(sample.meta$day)

readCounts = function(fname) {
  counts = readtsv(fname)
  # Fix gene names to remove the version number (e.g. ".4" in ENSG00000223972.4)
  gene_ids <- gsub("\\.[\\d]+", "", counts[,1], perl=T)
  counts[,1] = gene_ids
  rownames(counts) = gene_ids
  counts
}

counts = readCounts(counts.fname)
gene.meta = counts[,c(1,2)]
counts = counts[, -c(1,2)]
colnames(counts) = tolower(colnames(counts))
rownames(gene.meta) = gene.meta$gene_id
rownames(counts) = gene.meta$gene_id

readCounts.df = data.frame(readCount = colSums(counts), sampleID = colnames(counts))
sample.meta = sample.meta %>% dplyr::inner_join(readCounts.df, by="sampleID")
rownames(sample.meta) = sample.meta$sampleID

coldata = data.frame(sampleID = colnames(counts))
dds = DESeqDataSetFromMatrix(counts, coldata, ~0)
colnames(dds) = colnames(counts)
vst = varianceStabilizingTransformation(dds)
dds = estimateSizeFactors(dds)
#View(assay(vst))
counts.deseq = counts(dds, normalized=T)
colnames(counts.deseq) = colnames(counts)

rpm = apply(counts.deseq, MARGIN=2, FUN=function(x) x / sum(x)) * 1e6
rpkm = rpm * 1e3
rpkm = apply(rpkm, MARGIN=2, FUN=function(x) x / gene.meta$length)
geneMedianRpkm = apply(rpkm, MARGIN=1, FUN=median)
geneMeanRpkm = apply(rpkm, MARGIN=1, FUN=mean)
expressedGenes = names(geneMedianRpkm[geneMedianRpkm >= 1])

rpkm.df = cbind(data.frame(gene_id = rownames(rpkm)), rpkm)
rpkm.mat = as.matrix(rpkm.df[,-1])

txdb = readRDS(file.path(root, "reference/GRCh38/Homo_sapiens.GRCh38.89.transcript_data.rds"))
genedb = txdb %>% dplyr::select(ensembl_gene_id, external_gene_name, gene_biotype) %>% 
  unique() %>% dplyr::rename("gene_id" = "ensembl_gene_id", "gene_name" = "external_gene_name")
```

## Compare with Barres human brain cell types

Unlike the comparison with mouse brain cell types, the heatmap and correlations with human sorted brain cell types show that both BOB and Kolf2 neurons are much closer to fetal astrocytes than to any other brain cell type.

```{r Compare.Barres_rnaseq, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
barres.df = readtsv(file.path(root, "reference/Barres_RNAseq/barres_rnaseq.human.txt"))

# Remove duplicate gene IDs. First order by expression level so that if one version of the
# gene is expressed we capture that one
barres.df = barres.df[order(-rowMeans(barres.df[,-1])),]
barres.df = barres.df[!duplicated(barres.df$Gene),]
rownames(barres.df) = barres.df$Gene
barres.df = HGNCToEnsemblRownames(barres.df, txRefPath)

# To keep visualisation simple, for each cell type let's take the average
# RPKM of the samples of that type
barres.df = barres.df[,-1]
barres.meta = readtsv(file.path(root, "reference/Barres_RNAseq/barres_rnaseq.human.metadata.txt"))
rownames(barres.meta) = barres.meta$sampleID
# Make sure the metadata is in the same order as the cols of the matrix
barres.meta = barres.meta[colnames(barres.df),]
barres.meta$celltype = paste0("BAR_", barres.meta$celltype)
barres.avg.df = data.frame(gene_id = rownames(barres.df))
for (celltype in unique(barres.meta$celltype)) {
  barres.avg.df[, celltype] = rowMeans(barres.df[, barres.meta$celltype == celltype, drop=F])
}

# Check that none of the genes are completely zero, indicating that they probably
# weren't quantified at all
#sum(rowSums(barres.df[,-1]) == 0)

#rpkm.df = log2(rpkm+0.1) %>% as.data.frame()
rpkm.df = rpkm %>% as.data.frame()
rpkm.df$gene_id = rownames(rpkm.df)

rpkm.cmp = rpkm.df %>% dplyr::inner_join(barres.avg.df, by="gene_id") %>% dplyr::select(-gene_id)
colnames(rpkm.cmp)[1:13] = sample.meta[colnames(rpkm.cmp)[1:13],]$Sample_name
# Remove genes where our samples all have zeroes, indicating that these genes
# probably weren't quantified at all
sum(rowSums(rpkm.cmp[,1:13]) == 0) # 1187 genes
rpkm.cmp = rpkm.cmp[rowSums(rpkm.cmp[,1:13]) > 0,]

#cor.pearson = cor(log2(rpkm.cmp + 0.1))
#pheatmap(cor.pearson, main="Heatmap of gene expression correlations (Pearson)", fontsize_row = 9, fontsize_col = 8, #display_numbers = T, number_format = "%.2f", fontsize_number=8)

cor.spearman = cor(rpkm.cmp, method="spearman")
pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 9, fontsize_col = 8, display_numbers = T, number_format = "%.2f", fontsize_number=8)
```

Since the above was a strange result, I'm including the mouse data here in a large heatmap for comparison. This shows that mouse astrocytes are most similar to human astrocytes (mature and fetal), and mouse neurons are most similar to human fetal astrocytes followed by both mature astrocytes and neurons.

Surprisingly, mouse neurons are closer to our human iPSC-derived cells than to most other human sorted cell types from the Barres lab. I suspect that technical differences must be contributing here. It makes me wonder whether the human fetal astrocyte data from the Barres lab has technical differences with the remaining human sorted cell type data.

```{r Compare.Barres_rnaseq.2, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
barres.mouse.df = readtsv(file.path(root, "reference/Barres_RNAseq/barres_rnaseq.mouse_avg.txt")) %>%
  dplyr::rename(gene_name = Gene_symbol)
barres.mouse.df$gene_name = toupper(barres.mouse.df$gene_name)
colnames(barres.mouse.df)[3:ncol(barres.mouse.df)] = paste0("M_", colnames(barres.mouse.df)[3:ncol(barres.mouse.df)])

rpkm.cmp = rpkm.df %>% dplyr::inner_join(barres.avg.df, by="gene_name") %>%
  dplyr::inner_join(barres.mouse.df, by="gene_name") %>%
  dplyr::select(-gene_id, -gene_name, -Description)

colnames(rpkm.cmp)[1:13] = sample.meta[colnames(rpkm.cmp)[1:13],]$Sample_name

#cor.pearson = cor(log2(rpkm.cmp + 0.1))
#pheatmap(cor.pearson, main="Heatmap of gene expression correlations (Pearson)", fontsize_row = 9, fontsize_col = 8, display_numbers = T, number_format = "%.2f", fontsize_number=7)

cor.spearman = cor(rpkm.cmp, method="spearman")
pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 9, fontsize_col = 8, display_numbers = T, number_format = "%.2f", fontsize_number=7)
```

We do a PCA with axes defined by the Barres lab human samples, and project our samples onto these. In PC1 vs PC2, iNeurons are most similar to Barres neurons, while NPCs and NPC-derived neurons are more similar to Barres astrocytes. It's hard to know what higher PC plots are telling us.

```{r Compare.Barres_rnaseq.3, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
#source("~js29/src/R/pca.projection.R")
source("./pca.projection.R")

ref.mat = log2(as.matrix(barres.df)+1)
rownames(ref.mat) = barres_gene_names
ref.meta = data.frame(ID=colnames(ref.mat),
                      group=paste0("BAR_", barres.meta$celltype))

sample.mat = log2(rpkm.df[,1:13] %>% as.matrix() + 1)
rownames(sample.mat) = rpkm.df$gene_name
sample.pca.meta = data.frame(ID=sample.meta[colnames(sample.mat), ]$Sample_name,
                             group=paste0("IPS_", sample.meta[colnames(sample.mat), ]$celltype))
colnames(sample.mat) = sample.pca.meta$ID

pcaProj = pcaProjection(ref.mat, ref.meta, sample.mat, sample.pca.meta)
pca.s = summary(pcaProj$pcaRes)
groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC2")
groupLoc$group = rownames(groupLoc)
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.05*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange
ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC2, col=group)) +
  geom_point(alpha=0.7) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group, col=group), data=groupLoc, size=3) +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pca.s$importance["Proportion of Variance","PC1"]*100)) +
  ylab(sprintf("PC2 (%.1f%%)", pca.s$importance["Proportion of Variance","PC2"]*100))

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC3")
groupLoc$group = rownames(groupLoc)
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.05*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange
ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC3, col=group)) +
  geom_point(alpha=0.7) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group, col=group), data=groupLoc, size=3) +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pca.s$importance["Proportion of Variance","PC1"]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pca.s$importance["Proportion of Variance","PC3"]*100))

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC4")
groupLoc$group = rownames(groupLoc)
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.05*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange
ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC4, col=group)) +
  geom_point(alpha=0.7) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group, col=group), data=groupLoc, size=3) +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pca.s$importance["Proportion of Variance","PC1"]*100)) +
  ylab(sprintf("PC4 (%.1f%%)", pca.s$importance["Proportion of Variance","PC4"]*100))

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC2", "PC3")
groupLoc$group = rownames(groupLoc)
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.05*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange
ggplot(pcaProj$sampleCoords, aes(x=PC2, y=PC3, col=group)) +
  geom_point(alpha=0.7) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group, col=group), data=groupLoc, size=3) +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC2 (%.1f%%)", pca.s$importance["Proportion of Variance","PC2"]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pca.s$importance["Proportion of Variance","PC3"]*100))

```

Let's look at a heatmap of how similar the iPSC-derived samples are to GTEx tissues. All of the iPSC-derived samples are closer to brain than to other GTEx tissues. However, it's worth noting that iPSCs also have relatively high similarity with GTEx brain as well as other GTEx tissues.

```{r Compare.GTEx.heatmap.1, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
gtex.df = readRDS(gtex.rpkmfile)
gtex.meta.all = read.delim(gtex.metafile, stringsAsFactors = F)

gtexCols = seq(2, ncol(gtex.df))

# Find genes in common with GTEx samples
refGenes = data.frame(geneID = unique(as.character(rownames(gtex.df))))
sampleGenes = data.frame(geneID = unique(as.character(rownames(rpkm.mat))))
gtexGenesInCommon = refGenes %>% dplyr::inner_join(sampleGenes, by="geneID")

# Identify genes that are lowly expressed across all GTEx samples, as these
# aren't of interest. Reducing the size of the dataset will make this feasible
# to run.
gtex.mat = as.matrix(gtex.df[gtexGenesInCommon$geneID, gtexCols])
rm(gtex.df)
gc()
numSamplesExpressingGene.gtex = apply(gtex.mat, 1, FUN = function(x) sum(x > 0.1))
#hist(numSamplesExpressingGene.gtex, breaks=50)
#hist(numSamplesExpressingGene.gtex[numSamplesExpressingGene.gtex < 50], breaks=50)
#lowExprGenes = names(numSamplesExpressingGene.gtex[numSamplesExpressingGene.gtex < 50])

# Identify genes that are not expressed at all in our samples - these are
# probably technical differences and so not informative.
rpkm.mat.cmp = rpkm.mat[gtexGenesInCommon$geneID, ]
numSamplesExpressingGene.nongtex = apply(rpkm.mat.cmp, 1, FUN = function(x) sum(x > 0))

genesToUse = names(numSamplesExpressingGene.gtex[numSamplesExpressingGene.gtex >= 50 & numSamplesExpressingGene.nongtex > 0])
gtex.mat = gtex.mat[genesToUse,]
gc()

gtex.meta = gtex.meta.all[colnames(gtex.mat),]
gtex.avg.df = data.frame(gene_id = rownames(gtex.mat))
for (celltype in unique(gtex.meta$SMTS)) {
  gtex.avg.df[, celltype] = rowMeans(gtex.mat[, gtex.meta$SMTS == celltype, drop=F])
}

rownames(rpkm.df) = rpkm.df$gene_id
rpkm.df = rpkm.df[genesToUse,]
rpkm.cmp = rpkm.df %>% dplyr::inner_join(gtex.avg.df, by="gene_id") %>% dplyr::select(-gene_id, -gene_name)
colnames(rpkm.cmp)[1:13] = sample.meta[colnames(rpkm.cmp)[1:13],]$Sample_name

#cor.pearson = cor(log2(rpkm.cmp + 0.1))
#pheatmap(cor.pearson, main="Heatmap of gene expression correlations (Pearson)", fontsize_row = 9, fontsize_col = 8, #display_numbers = T, number_format = "%.2f", fontsize_number=8)

cor.spearman = cor(rpkm.cmp, method="spearman")
pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 8, fontsize_col = 7, display_numbers = T, number_format = "%.2f", fontsize_number=5)

```

We can also include the Barres lab samples in the same heatmap with GTEx. To make visualization feasible I use average expression for our IPSC-derived iNeurons, NPCs, and neurons. Note that fewer genes are used in this comparison, because we only use those that were also quantified in the Barres lab samples (18,261 now, vs. 37820 in common between our iPSC-derived samples and GTEx).

```{r Compare.GTEx.heatmap.2, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
# Add the Barres lab samples to the current set of samples. First we want to
# simplify things by taking the average of each group of our samples
rpkm.avg.df = data.frame(gene_name = rpkm.df$gene_name, gene_id = rpkm.df$gene_id)
rpkm.tmp = rpkm.df %>% dplyr::select(-gene_name, -gene_id)
celltypes = paste0("IPS_", sample.meta$celltype)
for (celltype in unique(celltypes)) {
  rpkm.avg.df[, celltype] = rowMeans(rpkm.tmp[, celltypes == celltype, drop=F])
}
rm(rpkm.tmp)

# Merge with Barres lab samples
rpkm.cmp = rpkm.avg.df %>% dplyr::inner_join(barres.avg.df, by="gene_name") %>%
  dplyr::inner_join(gtex.avg.df, by="gene_id")
genesInCommonBarresGTExOurs = rpkm.cmp$gene_id
genesInCommonBarresGTExOurs.hgnc = rpkm.cmp$gene_name
rpkm.cmp = rpkm.cmp %>% dplyr::select(-gene_id, -gene_name)

cor.spearman = cor(rpkm.cmp, method="spearman")
pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 8, fontsize_col = 7, display_numbers = T, number_format = "%.2f", fontsize_number=5)
```

Let's see where the samples project on a PCA of GTEx samples. PC1 vs PC2 separates GTEx brain and blood from other samples, and the iPSC-derived neurons don't cluster with brain. PC1 vs PC3 doesn't tell us much. In PC2 vs PC3, iPSC-derived neurons cluster directly overlapping GTEx brain, with blood and muscle being separate.
In the end, I'm not sure that the GTEx data tell us much. The sorted cell data seems much more informative.

```{r Compare.GTEx.PCA.1, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
gtexNsamples = sum(grepl("GTEX", colnames(gtex.mat)))
gtexSelect = 1
gtexCols = seq(1, gtexNsamples, gtexSelect)

# For the PCA we only want to use GTEx as reference samples
gtex.rpkm.log.mat = as.matrix(log2(gtex.mat[genesToUse, gtexCols] + 1))
#saveRDS(gtex.rpkm.log.mat, file=paste0(outputPath, "gtex.rpkm.log.mat.rds"))

gtex.meta$SMTS[gtex.meta$SMTS == "IPSDSN P1"] = "IPSDSN"
gtex.meta$SMTS[gtex.meta$SMTS == "IPSDSN P2"] = "IPSDSN"
otherSamples.meta = gtex.meta[gtex.meta$SMTS %in% c("IPSC", "DRG", "IPSDSN"), ]
otherSamples.mat = as.matrix(log2(gtex.mat[genesToUse, otherSamples.meta$SAMPID] + 1))
rm(gtex.mat)
gc()

rownames(sample.mat) = rpkm.df$gene_id
sample.pca.mat.combined = cbind(sample.mat[genesToUse,], otherSamples.mat)
sample.pca.meta.combined = rbind(sample.pca.meta, data.frame(ID=otherSamples.meta$SAMPID, group=otherSamples.meta$SMTS))

gtex.pca.meta = gtex.meta[grepl("GTEX", gtex.meta$SAMPID), ]
gtex.pca.meta$ID = gtex.pca.meta$SAMPID
gtex.pca.meta$group = gtex.pca.meta$SMTS
pcaProj = bigpcaProjection(gtex.rpkm.log.mat, gtex.pca.meta, sample.pca.mat.combined, sample.pca.meta.combined)
#saveRDS(pcaProj, file=paste0(outputPath, "bigPCA.result.rds"))
#pcaProj = readRDS(paste0(outputPath, "bigPCA.result.rds"))

ourSampleCoords = pcaProj$sampleCoords[pcaProj$sampleCoords$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]

pcaVarExp = pcaProj$pcaRes$Evalues / sum(pcaProj$pcaRes$Evalues)
groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC2")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.1*xRange
xlim2 = max(groupLoc$x) + 0.1*xRange


p1 = ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC2)) +
  geom_point(aes(x=PC1, y=PC2, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC1, y=PC2, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pcaVarExp[1]*100)) +
  ylab(sprintf("PC2 (%.1f%%)", pcaVarExp[2]*100))
print(p1)

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC3")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.1*xRange
xlim2 = max(groupLoc$x) + 0.1*xRange
p2 = ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC3)) +
  geom_point(aes(x=PC1, y=PC3, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC1, y=PC3, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pcaVarExp[1]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pcaVarExp[3]*100))
print(p2)

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC2", "PC3")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.1*xRange
xlim2 = max(groupLoc$x) + 0.1*xRange
p3 = ggplot(pcaProj$sampleCoords, aes(x=PC2, y=PC3)) +
  geom_point(aes(x=PC2, y=PC3, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC2, y=PC3, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC2 (%.1f%%)", pcaVarExp[2]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pcaVarExp[3]*100))
print(p3)

if (saveFiles) {
  pdf(file.path(outputPath, "ipsneurons.PCA.gtex.pdf"), width=12, height=8)
  pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 8, fontsize_col = 7, display_numbers = T, number_format = "%.2f", fontsize_number=5)
  print(p1)
  print(p2)
  print(p3)
  dev.off()
}
```

In the end, I'm not sure that the GTEx data tell us much. The sorted cell data seems much more informative.

Let's add in the Barres lab samples and see where they project. Again, the overall PCA plot may be different since we only use genes in common between all three sets of samples (ours, Barres lab, GTEx).

```{r Compare.GTEx.PCA.2, warning=FALSE, message=FALSE, echo=FALSE, fig.width=10, fig.height=8}
# For the PCA we only want to use GTEx as reference samples
gtex.rpkm.log.mat2 = gtex.rpkm.log.mat[genesInCommonBarresGTExOurs,]
otherSamples.mat2 = otherSamples.mat[genesInCommonBarresGTExOurs,]
barresSamples.mat = log2(as.matrix(barres.df[genesInCommonBarresGTExOurs,]+1))

gc()

sample.pca.mat.combined = cbind(sample.mat[genesInCommonBarresGTExOurs,], otherSamples.mat2, barresSamples.mat)
sample.pca.meta.combined = rbind(sample.pca.meta,
                                 data.frame(ID=otherSamples.meta$SAMPID, group=otherSamples.meta$SMTS),
                                 data.frame(ID=barres.meta$sampleID, group=barres.meta$celltype))

pcaProj = bigpcaProjection(gtex.rpkm.log.mat2, gtex.pca.meta, sample.pca.mat.combined, sample.pca.meta.combined)
#saveRDS(pcaProj, file=paste0(outputPath, "bigPCA.result.rds"))
#pcaProj = readRDS(paste0(outputPath, "bigPCA.result.rds"))

#ourSampleCoords = pcaProj$sampleCoords[pcaProj$sampleCoords$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
ourSampleCoords = pcaProj$sampleCoords[grepl("IPS_iNeuron|IPS_NPC|IPS_neuron|BAR_*", pcaProj$sampleCoords$group), ]

pcaVarExp = pcaProj$pcaRes$Evalues / sum(pcaProj$pcaRes$Evalues)
groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC2")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.15*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange


p1 = ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC2)) +
  geom_point(aes(x=PC1, y=PC2, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC1, y=PC2, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pcaVarExp[1]*100)) +
  ylab(sprintf("PC2 (%.1f%%)", pcaVarExp[2]*100))
print(p1)

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC1", "PC3")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.15*xRange
xlim2 = max(groupLoc$x) + 0.05*xRange
p2 = ggplot(pcaProj$sampleCoords, aes(x=PC1, y=PC3)) +
  geom_point(aes(x=PC1, y=PC3, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC1, y=PC3, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC1 (%.1f%%)", pcaVarExp[1]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pcaVarExp[3]*100))
print(p2)

groupLoc = pcaGroupMeans(pcaProj$sampleCoords, "PC2", "PC3")
groupLoc$group = rownames(groupLoc)
ourGroupLoc = groupLoc[groupLoc$group %in% c("IPS_iNeuron", "IPS_NPC", "IPS_neuron"), ]
xRange = max(groupLoc$x) - min(groupLoc$x)
xlim1 = min(groupLoc$x) - 0.05*xRange
xlim2 = max(groupLoc$x) + 0.15*xRange
p3 = ggplot(pcaProj$sampleCoords, aes(x=PC2, y=PC3)) +
  geom_point(aes(x=PC2, y=PC3, col=group), alpha=0.7) + theme_bw(12) +
  geom_point(aes(x=PC2, y=PC3, col=group), data=ourSampleCoords, size=3) + theme_bw(12) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=groupLoc, size=3) +
  geom_text(mapping=aes(x=x, y=y, label=group), data=ourGroupLoc, size=3, fontface="bold") +
  scale_colour_discrete(guide = FALSE) +
  coord_cartesian(xlim=c(xlim1, xlim2)) +
  xlab(sprintf("PC2 (%.1f%%)", pcaVarExp[2]*100)) +
  ylab(sprintf("PC3 (%.1f%%)", pcaVarExp[3]*100))
print(p3)

if (saveFiles) {
  pdf(file.path(outputPath, "ipsneurons.PCA.gtex.pdf"), width=12, height=8)
  pheatmap(cor.spearman, main="Heatmap of gene expression correlations (Spearman)", fontsize_row = 8, fontsize_col = 7, display_numbers = T, number_format = "%.2f", fontsize_number=5)
  print(p1)
  print(p2)
  print(p3)
  dev.off()
}
```

